### PR TITLE
kernel: work: Add Missing args ASSERT

### PR DESCRIPTION
**Pull request by LingaoM**
1. Add missing assert for arguments
2. Remove meanless foreach loop
3. fix usage.

### CODE REVIEW

File: kernel/work.c [hunk 1]
```diff
@@ -974,10 +966,10 @@ static int cancel_delayable_async_locked(struct k_work_delayable *dwork)
 	return cancel_async_locked(&dwork->work);
 }
 
-int k_work_schedule_for_queue(struct k_work_q *queue,
-			       struct k_work_delayable *dwork,
-			       k_timeout_t delay)
+int k_work_schedule_for_queue(struct k_work_q *queue, struct k_work_delayable *dwork,
+			      k_timeout_t delay)
```
Comment (cfriedt) [inline kernel/work.c:970]: Please move formatting changes to a separate commit, if they are necessary.

File: kernel/work.c [hunk 2]
```diff
@@ -1010,10 +1001,10 @@ int k_work_schedule(struct k_work_delayable *dwork,
 	return ret;
 }
 
-int k_work_reschedule_for_queue(struct k_work_q *queue,
-				 struct k_work_delayable *dwork,
-				 k_timeout_t delay)
+int k_work_reschedule_for_queue(struct k_work_q *queue, struct k_work_delayable *dwork,
+				k_timeout_t delay)
```
Comment (cfriedt) [inline kernel/work.c:1005]: Please move formatting changes to a separate commit, if they are necessary.

File: kernel/work.c [hunk 3]
```diff
@@ -998,8 +990,7 @@ int k_work_schedule_for_queue(struct k_work_q *queue,
 	return ret;
 }
 
-int k_work_schedule(struct k_work_delayable *dwork,
-				   k_timeout_t delay)
+int k_work_schedule(struct k_work_delayable *dwork, k_timeout_t delay)
```
Comment (cfriedt) [inline kernel/work.c:993]: Please move formatting changes to a separate commit, if they are necessary.

File: kernel/work.c [hunk 4]
```diff
@@ -1034,8 +1025,7 @@ int k_work_reschedule_for_queue(struct k_work_q *queue,
 	return ret;
 }
 
-int k_work_reschedule(struct k_work_delayable *dwork,
-				     k_timeout_t delay)
+int k_work_reschedule(struct k_work_delayable *dwork, k_timeout_t delay)
```
Comment (cfriedt) [inline kernel/work.c:1028]: Please move formatting changes to a separate commit, if they are necessary.

**Review by cfriedt — COMMENTED** [review]
Looks ok to me, aside from seemingly unnecessary formatting changes

**Review by cfriedt — APPROVED** [review]
Still looks OK to me - I see something that is possibly `O(n^2)` being reduced to `O(n)` which seems good.

Just for my own clarification, I wanted to find where the `K_WORK_QUEUED` flag was being set or cleared.

```shell
$ grep -Rn "K_WORK_QUEUED" kernel/
kernel/work.c:221:      if (flag_test_and_clear(&work->flags, K_WORK_QUEUED_BIT)) {
kernel/work.c:335:      } else if (!flag_test(&work->flags, K_WORK_QUEUED_BIT)) {
kernel/work.c:360:                      flag_set(&work->flags, K_WORK_QUEUED_BIT);
kernel/work.c:451:                         & (K_WORK_QUEUED | K_WORK_RUNNING)) != 0U;
kernel/work.c:639:                      flag_clear(&work->flags, K_WORK_QUEUED_BIT);
```

In the first case, we perform another linear search through an slist. One thing to keep in mind is that an slist has the same storage requirements as a dlist, and a dlist item can be removed in `O(1)`, whereas an slist is `O(n)`. Maybe we want to switch to dlists for work queues?

Those are the only useful holes i can poke, mainly just hoping to draw attention.

LGTM

**Review by cfriedt — APPROVED** [review]
Still looks OK to me - I see something that is possibly `O(n^2)` being reduced to `O(n)` which seems good.

Just for my own clarification, I wanted to find where the `K_WORK_QUEUED` flag was being set or cleared.

```shell
$ grep -Rn "K_WORK_QUEUED" kernel/
kernel/work.c:221:      if (flag_test_and_clear(&work->flags, K_WORK_QUEUED_BIT)) {
kernel/work.c:335:      } else if (!flag_test(&work->flags, K_WORK_QUEUED_BIT)) {
kernel/work.c:360:                      flag_set(&work->flags, K_WORK_QUEUED_BIT);
kernel/work.c:451:                         & (K_WORK_QUEUED | K_WORK_RUNNING)) != 0U;
kernel/work.c:639:                      flag_clear(&work->flags, K_WORK_QUEUED_BIT);
```

In the first case, we perform another linear search through an slist. One thing to keep in mind is that an slist has the same storage requirements as a dlist, and a dlist item can be removed in `O(1)`, whereas an slist is `O(n)`. Maybe we want to switch to dlists for work queues?

Those are the only useful holes i can poke, mainly just hoping to draw attention.

LGTM

**Comment by LingaoM** [comment]
@peter-mitsis  Could you please a look? thanks.

**Review by peter-mitsis — APPROVED** [review]
